File: src/Restaurant/InMemory Entity Adapter/FTGO.Restaurant.InMemoryEntities/Entry.fs
namespace FTGO.Restaurant.InMemoryEntities

open System

[<CustomEquality; NoComparison>]
type internal Entry<'value> =
    {
        Version : uint64
        Deleted : bool
        Value : 'value
    }

    interface IEquatable<Entry<'value>> with
        member this.Equals other =
            if this.Deleted then
                false
            else
                this.Version = other.Version

    override this.Equals obj =
        match obj with
        | :? Entry<'value> as other ->
            (this :> IEquatable<_>).Equals other
        | _ -> false

    override this.GetHashCode () =
        HashCode.Combine (this.Deleted, this.Version)


module internal Entry =

    let create version value = {
        Version = version
        Deleted = false
        Value = value
    }

    let compare version = {
        Version = version
        Deleted = false
        Value = Unchecked.defaultof<_>
    }

    let delete () = {
        Version = Unchecked.defaultof<_>
        Deleted = true
        Value = Unchecked.defaultof<_>
    }

File: src/AD/AD.InMemoryStore/Entry.cs
namespace AD.InMemoryStore
{
    class Entry<TValue>
    {
        public static Entry<TValue> Create(TValue value, Version version) => new() { Value = value, Version = version };
        public static Entry<TValue> Compare(Version version) => new() { Version = version };
        public static Entry<TValue> Delete() => new() { Version = Version.Deleted };

        private Entry() { }

        public TValue Value { get; private init; } = default!;
        public Version Version { get; private init; }

        public bool Deleted => Version.Value == Version.Deleted.Value;

        public override bool Equals(object? obj) =>
            !Deleted && obj is Entry<TValue> other && Version.Value == other.Version.Value;

        public override int GetHashCode() =>
            Version.GetHashCode();

        public static implicit operator (TValue, Version)(Entry<TValue> entry) =>
            (entry.Value, entry.Version);
    }
}
